Sources/diag.swift

```diff
@@ -3,24 +3,102 @@
  *
  *  This file provides functionality for creating or retrieving the diagonal of a matrix.
  *
- *  Author: Philip Erickson
- *  Creation Date: 26 Dec 2016
+ *  Author: Tor Rafsol Løseth
+ *  Creation Date: 07 Jul 2018
  *
  *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  *  in compliance with the License. You may obtain a copy of the License at
  *
  *      http://www.apache.org/licenses/LICENSE-2.0
  *
- *  Unless required by applicable law or agreed to in writing, software distributed under the 
- *  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either 
- *  express or implied. See the License for the specific language governing permissions and 
+ *  Unless required by applicable law or agreed to in writing, software distributed under the
+ *  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing permissions and
  *  limitations under the License.
  *
- *  Copyright 2016 Philip Erickson
+ *  Copyright 2018 Tor Rafsol Løseth
  **************************************************************************************************/
 
- public func diag() -> ()
- {
-     // FIXME: implement this
-     fatalError("Not yet implemented")
- }
\ No newline at end of file
+
+/// Returns a square diagonal matrix with the elements of vector v on the main diagonal,
+/// and zeros everywhere else.
+///
+/// - Parameter v   : A vector data structure
+/// - Returns: A matrix data structure
+public func diag(v: Vector<Double>) -> Matrix<Double>
+{
+    return diag(v: v, k: 0)
+}
+
+
+/// Returns a square diagonal matrix with the elements of vector v on the kth diagonal,
+/// and zeros everywhere else.
+///
+/// - Parameters:
+///   - v: A vector data structure.
+///   - k: The kth diagonal, k = 0 is the main diagonal, k > 0 is above the main diagonal, and k < 0 is below the main diagonal.
+/// - Returns: A square matrix data structure.
+public func diag(v: Vector<Double>, k: Int) -> Matrix<Double>
+{
+    
+    let length = v.count + abs(k)
+    let A = Matrix(length, length, value: 0.0)
+    var data = A.data
+    
+    // Diagonals have indices intervals at every other:
+    let interval = A.rows + 1
+    
+    let diagonalStartIndex = k >= 0 ? k : length * abs(k)
+    
+    let vectorIndices = (0..<v.count).map {index in
```

Comment:
My one thought is by precomputing these, we allocate a `n` size array when we really only need each vector index once. We could pull the data assignment from the for loop below into this loop (and change this loop from a map to a for) and avoid this. Seems like it doesn't matter much though.